shared/engine/index.js

```diff
@@ -1,435 +1,284 @@
-      cancel: () => {
-        this.logLocal('Cancelling response', method, seqid, param)
-        cleanup()
-        once = true
-      },
-      result: (...args: Array<any>) => {
-        if (once) {
-          if (printRPC) {
-            this.logLocal('RPC ▼ result bailing on additional calls: ', method, seqid, param, ...args)
-          }
-          return
-        }
-        once = true
-
-        if (printRPC) {
-          this.logLocal('RPC ▼ result: ', method, param, ...args)
-        }
-        if (response) {
-          cleanup()
-          response.result(...args)
-        } else if (__DEV__) {
-          console.warn('Calling response.result on non-response object: ', method)
-        }
-      },
-      error: (...args: Array<any>) => {
-        if (once) {
-          if (printRPC) {
-            this.logLocal('RPC ▼ error bailing on additional calls: ', method, seqid, param, ...args)
-          }
-          return
-        }
-        once = true
+    console.warn(`Unknown incoming rpc: ${sessionID} ${method}`)
 
-        if (printRPC) {
-          this.logLocal('RPC ▼ error: ', method, param, ...args)
-        }
-
-        if (response) {
-          cleanup()
-          response.error(...args)
-        } else if (__DEV__) {
-          console.warn('Calling response.error on non-response object: ', method)
-        }
-      },
+    if (__DEV__ && this._failOnError) {
+      throw new Error(`unhandled incoming rpc: ${sessionID} ${method} ${JSON.stringify(param)}${response ? '. has response' : ''}`)
     }
 
-    if (__DEV__ && KEYBASE_RPC_DELAY) {
-      const result = wrappedResponse.result
-      const error = wrappedResponse.error
-      wrappedResponse.result = (...args) => {
-        console.log('RPC Delayed')
-        setTimeout(() => { result(...args) }, KEYBASE_RPC_DELAY)
-      }
-      wrappedResponse.error = (...args) => {
-        console.log('RPC Delayed')
-        setTimeout(() => { error(...args) }, KEYBASE_RPC_DELAY)
-      }
-    }
+    response && response.error && response.error({
+      code: ConstantsStatusCode.scgeneric,
+      desc: `Unhandled incoming RPC ${sessionID} ${method}`,
+    })
+  }
 
-    // Incoming calls have no sessionID so let's ignore
-    if (sessionID) {
-      this.sessionIDToResponse[String(sessionID)] = wrappedResponse
+  // An incoming rpc call
+  _rpcIncoming (payload: {method: MethodKey, param: Array<Object>, response: ?Object}) {
+    const {method, param: incomingParam, response} = payload
+    const param = incomingParam && incomingParam.length ? incomingParam[0] : {}
+    const {seqid, cancelled} = response || {seqid: 0, cancelled: false}
+    const {sessionID} = param
 
-      if (printOutstandingRPCs) {
-        this.responseMeta[sessionID] = {method, param}
+    if (cancelled) {
+      this._handleCancel(seqid)
+    } else {
+      const session = this._sessionsMap[String(sessionID)]
+      if (session && session.incomingCall(method, param, response)) { // Part of a session?
```

Comment:
Just a note for myself and other reviewers:

This lets the session say "I can't handle this" even if there is a sesssion that matches that session id.

Useful for stuff like pinentry where we don't want each session handling it. It's okay to just fallback